internal/cmd/controller/helmops/reconciler/fleethelm_controller.go

```diff
@@ -0,0 +1,413 @@
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// HelmAppReconciler reconciles a HelmApp resource to create and apply bundles for helm charts
+type HelmAppReconciler struct {
+	client.Client
+	Scheme    *runtime.Scheme
+	Scheduler quartz.Scheduler
+	Workers   int
+	ShardID   string
+	Recorder  record.EventRecorder
+}
+
+func (r *HelmAppReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&fleet.HelmApp{},
+			builder.WithPredicates(
+				predicate.Or(
+					predicate.GenerationChangedPredicate{},
+					predicate.AnnotationChangedPredicate{},
+					predicate.LabelChangedPredicate{},
+				),
+			),
+		).
+		WithEventFilter(sharding.FilterByShardID(r.ShardID)).
+		WithOptions(controller.Options{MaxConcurrentReconciles: r.Workers}).
+		Complete(r)
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// The Reconcile function compares the state specified by
+// the GitRepo object against the actual cluster state, and then
+// performs operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.15.0/pkg/reconcile
+func (r *HelmAppReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	if !experimentalHelmOpsEnabled() {
+		return ctrl.Result{}, nil
+	}
+	logger := log.FromContext(ctx).WithName("HelmApp")
+	helmapp := &fleet.HelmApp{}
+
+	if err := r.Get(ctx, req.NamespacedName, helmapp); err != nil && !errors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	} else if errors.IsNotFound(err) {
+		return ctrl.Result{}, nil
+	}
+
+	// Finalizer handling
+	purgeBundlesFn := func() error {
+		nsName := types.NamespacedName{Name: helmapp.Name, Namespace: helmapp.Namespace}
+		if err := finalize.PurgeBundles(ctx, r.Client, nsName, fleet.HelmAppLabel); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	if !helmapp.GetDeletionTimestamp().IsZero() {
+		err := purgeBundlesFn()
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if controllerutil.ContainsFinalizer(helmapp, finalize.HelmAppFinalizer) {
+			if err := deleteFinalizer(ctx, r.Client, helmapp, finalize.HelmAppFinalizer); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	if !controllerutil.ContainsFinalizer(helmapp, finalize.HelmAppFinalizer) {
+		if err := addFinalizer(ctx, r.Client, helmapp, finalize.HelmAppFinalizer); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Reconciling
+	logger = logger.WithValues("generation", helmapp.Generation, "chart", helmapp.Spec.Helm.Chart)
+	ctx = log.IntoContext(ctx, logger)
+
+	logger.V(1).Info("Reconciling HelmApp")
+
+	if helmapp.Spec.Helm.Chart == "" {
+		return ctrl.Result{}, nil
```

Comment:
I didn't because the equivalent in the `GitRepo`'s reconciles doesn't:

```go
	if gitrepo.Spec.Repo == "" {
		return ctrl.Result{}, nil
	}
```

I'm fine adding it, but I would add it in both controllers. (Maybe in another PR?)